dependencies.js

```diff
@@ -0,0 +1,412 @@
+				deps.forEach(function (depId) {
+					if (!(depId in entryMap)) {
+						throw new Error(id + ' depends on an unknown component ' + depId);
+					}
+
+					addToMap(depId, stack);
+					dependencies[depId] = true;
+					for (var transitiveDepId in map[depId]) {
+						dependencies[transitiveDepId] = true;
+					}
+				});
+			}
+
+			map[id] = dependencies;
+
+			stack.pop();
+		}
+
+		for (var id in entryMap) {
+			addToMap(id, []);
+		}
+
+		return map;
+	}
+
+	/**
+	 * Returns a function which resolves the aliases of its given id of alias.
+	 *
+	 * @param {EntryMap} entryMap
+	 * @returns {(idOrAlias: string) => string}
+	 */
+	function createAliasResolver(entryMap) {
+		/** @type {Object<string, string>} */
+		var map = {};
+
+		for (var id in entryMap) {
+			var entry = entryMap[id];
+			var aliases = toArray(entry && entry.alias);
+			aliases.forEach(function (alias) {
+				if (alias in map) {
+					throw new Error(alias + ' cannot be alias for both ' + id + ' and ' + map[alias]);
+				}
+				if (alias in entryMap) {
+					throw new Error(alias + ' cannot be alias of ' + id + ' because it is a component.');
+				}
+				map[alias] = id;
+			});
+		}
+
+		return function (idOrAlias) {
+			return map[idOrAlias] || idOrAlias;
+		};
+	}
+
+	/**
+	 * Creates an implicit DAG from the given components and dependencies and call the given `loadComponent` for each
+	 * component in topological order.
+	 *
+	 * @param {DependencyMap} dependencyMap
+	 * @param {StringSet} ids
+	 * @param {(id: string) => T} loadComponent
+	 * @param {(before: T, after: T) => T} series
+	 * @param {(values: T[]) => T} parallel
+	 * @returns {T}
+	 * @template T
+	 */
+	function loadComponentsInOrder(dependencyMap, ids, loadComponent, series, parallel) {
+		/** @type {Object<string, T>} */
+		var cache = {};
+
+		/**
+		 * A set of ids of nodes which are not depended upon by any other node in the graph.
+		 * @type {StringSet}
+		 */
+		var ends = {};
+
+		/**
+		 * Loads the given component and its dependencies or returns the cached value.
+		 *
+		 * @param {string} id
+		 * @returns {T}
+		 */
+		function handleId(id) {
+			if (id in cache) {
+				return cache[id];
+			}
+
+			// assume that it's an end
+			// if it isn't, it will be removed later
+			ends[id] = true;
+
+			// all dependencies of the component in the given ids
+			var dependsOn = [];
+			for (var depId in dependencyMap[id]) {
+				if (depId in ids) {
+					dependsOn.push(depId);
+				}
+			}
+
+			/**
+			 * The value to be returned.
+			 * @type {T}
+			 */
+			var value;
+
+			if (dependsOn.length === 0) {
+				value = loadComponent(id);
+			} else {
+				var depsValue = parallel(dependsOn.map(function (depId) {
+					var value = handleId(depId);
+					// none of the dependencies can be ends
+					delete ends[depId];
+					return value;
+				}));
+				value = series(depsValue, loadComponent(id));
```

Comment:
Yeah, this won't ensure all the `depsValue` promises are resolved before kicking off the request to the `id`.

Reply:
Fixed.